agent/src/main/java/com/teamscale/jacoco/agent/store/upload/UploadStoreBase.java

```diff
@@ -94,38 +92,38 @@ protected boolean tryUpload(String xml) {
 		} catch (IOException e) {
 			logger.error("Failed to upload coverage to {}. Probably a network problem", uploadUrl, e);
 			return false;
-		} catch (UploadStoreException e) {
+		} catch (UploaderException e) {
 			logger.error("Failed to upload coverage to {}. The configuration is probably incorrect", uploadUrl, e);
 			return false;
 		}
 	}
 
 	/**
-	 * Creates the zip file to upload which includes the given coverage XML and all
-	 * {@link #additionalMetaDataFiles}.
+	 * Creates the zip file to upload which includes the given coverage XML and all {@link #additionalMetaDataFiles}.
 	 */
-	private byte[] createZipFile(String xml) throws IOException {
-		try (ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream()) {
-			try (ZipOutputStream zipOutputStream = new ZipOutputStream(byteArrayOutputStream);
-				 OutputStreamWriter writer = new OutputStreamWriter(zipOutputStream, "UTF-8")) {
-				fillZipFile(zipOutputStream, writer, xml);
+	private File createZipFile(File coverageFile) throws IOException {
+		File zipFile = createZipFileObject(coverageFile);
+		try (FileOutputStream fileOutputStream = new FileOutputStream(zipFile)) {
+			try (ZipOutputStream zipOutputStream = new ZipOutputStream(fileOutputStream)) {
+				fillZipFile(zipOutputStream, coverageFile);
 			}
-			return byteArrayOutputStream.toByteArray();
+			return zipFile;
 		}
 	}
 
+	private File createZipFileObject(File coverageFile) {
+		Path parentFolder = coverageFile.getParentFile().toPath();
+		String filename = coverageFile.getName().substring(0, coverageFile.getName().length() - 4) + ".zip";
+		return new File(parentFolder.resolve(filename).toString());
+	}
+
 	/**
-	 * Fills the upload zip file with the given coverage XML and all
-	 * {@link #additionalMetaDataFiles}.
+	 * Fills the upload zip file with the given coverage XML and all {@link #additionalMetaDataFiles}.
 	 */
-	private void fillZipFile(ZipOutputStream zipOutputStream, OutputStreamWriter writer, String xml)
+	private void fillZipFile(ZipOutputStream zipOutputStream, File coverageFile)
 			throws IOException {
 		zipOutputStream.putNextEntry(new ZipEntry("coverage.xml"));
-		writer.write(xml);
-
-		// We flush the writer, but don't close it here, because closing the writer
-		// would also close the zipOutputStream, making further writes impossible.
-		writer.flush();
+		zipOutputStream.write(FileSystemUtils.readFileBinary(coverageFile));
```

Comment:
this again reads in the entire XML file as a byte array. Isn't that what we want to avoid at all costs, since the XMLs can be big? Maybe rather use `FileSystemUtils.copy` to copy the data over in chunks?

It's very easy to make this mistake. Should we create an abstraction `CoverageFile` that encapsulates and hides a private `Path` to a coverage file on disk and only offers a `copy` method to copy the file contents in this way into an output stream? That way, future maintainers are forced to think about this memory consumption issue and can't just accidentally read in the entire large file into memory.

Reply:
I like the idea. Unfortunately, we need a `java.io.Filo` to create the request body for the Teamscale upload, for example. So we need to expose the File anyways. I'll create a `CoverageFile` class anyway, as it makes it at least a little bit harder to do the wrong thing and we can advise against using the getter for the `java.io.File` in the documentation.